### PR TITLE
chore: WebKit support (development-only)

### LINKED ISSUE #6422: Safari (WebKit) Support
> **Current Status** - _last updated August 4, 2022_
> 
> ✅ Prioritized and on the roadmap
> ✅ Doing exploratory work to assess requirements and timelines
> ✅ Under active development
> ℹ️ _Cypress currently supports [Firefox](https://cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/) and Chrome-family browsers (including Edge and Electron)_.

**Objective**
---
Add support for running Cypress tests within WebKit, which is the engine used by the [Safari](https://www.apple.com/safari/) web browser.

**Task List**
---

- [x] Initial PR to add development-only support: https://github.com/cypress-io/cypress/pull/15533
- [ ] Fix required driver/other functionality
	- [x] Network functionality https://github.com/cypress-io/cypress/pull/23232
	- [x] Keyboard tests (`type*.cy.js`) https://github.com/cypress-io/cypress/pull/23522
	- [x] Input tests (`click.cy.js`, focus tests...) https://github.com/cypress-io/cypress/pull/23522
	- [x] Stack traces https://github.com/cypress-io/cypress/pull/23730
	- [x] Check component testing
	- [x] Video recording #23579 
	- [x] Handle downloads https://github.com/cypress-io/cypress/pull/23662
	- [x] Fix global mode detection of npm module https://github.com/cypress-io/cypress/pull/23711
	- [ ] "Network" GraphQL error fix - wip
	- [x] [Fix open mode close](https://github.com/cypress-io/cypress/pull/15533#issuecomment-1211506734) https://github.com/cypress-io/cypress/pull/23662
	- [x] Disable `forceNetworkError` https://github.com/cypress-io/cypress/pull/23711
	- [x] Fix cross-origin unhandledRejection events: https://github.com/cypress-io/cypress/pull/23588
- [ ] Possibly fix less critical functionality
	- [ ] Multi-domain implementation https://github.com/cypress-io/cypress/pull/23442
	- [ ] Fix websocket polling/proxying/allowedSockets
	- [ ] [Fix devicePixelRatio](https://github.com/cypress-io/cypress/pull/15533#issuecomment-1210169765)
	- [x] Fix/skip focus from app
- [x] Documentation - limitations, Linux dependencies https://github.com/cypress-io/cypress-documentation/pull/4710
- [x] Add experimental flag, make available in production, update public types https://github.com/cypress-io/cypress/pull/23711

### PR DESCRIPTION
**Pull request by flotwig**
- For #6422 

### User facing changelog

n/a - does not expose WebKit in production yet, only in development

### Additional details

- This PR adds development-only support for the WebKit browser. It will detect `playwright-webkit` installed in a project and if it is found, it will be made available in Cypress as `--browser webkit`
- See https://github.com/cypress-io/cypress/issues/6422 for remaining work - trying to keep PRs in this epic small since there's a lot to parse

Pre-merge tasks:

- [x] Merge with latest develop
- [x] Add driver test CI job
- [x] Double-check that --headless works w/o Xvfb

### How has the user experience changed?

In development only, WebKit will be detected in run/open mode if `playwright-webkit` is installed. 

image

### PR Tasks

- [x] Have tests been added/updated?
	- Only tests relating to browser detection have been added/updated. Future test updates will come as remaining functionality is fixed, see #6422 for details.
- [na] Has the original issue or this PR been tagged with a release in ZenHub? 
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
	- Public type changes will come in the last PR for experimentalWebKitSupport, see #6422

Co-authored-by: Weyert de Boer <weyert@innerfuse.biz>

### CODE REVIEW

File: packages/app/src/runner/index.ts
Comment (tbiethman) [inline packages/app/src/runner/index.ts:42]: Thanks for the helpful reference here

**Comment by lmiller1990** [comment]
Still going over the code, but I took it for a test drive. It functions like it should, but I am noticing everything in the webkit driver is kind of blurry.

Compare the text from GH (left) and the spec list (right). I am guessing this isn't on our end, but the webkit-playwright code? Either way, users might notice this - we might want to include a note in the experimental release (unless the bug is localised to my machine?)

I am on an M1 running under arm64 if that is relevant.

I will continue the code review tomorrow. I am excited for WebKit support!

image

**Review by lmiller1990 — COMMENTED** [review]
I'm still working through the code - posted my finding so far, but functionality seems to work as expected.

**Comment by tbiethman** [comment]
@lmiller1990 I can confirm I see the same blurry rendering on my Macbook display, but it looks crisp on my external display (with a much lower pixel density).

**Comment by flotwig** [comment]
@lmiller1990 @tbiethman We check and set 2x pixel density on other browsers but not in WebKit. I can add it to the list of things to fix in the parent epic, but in the spirit of keeping this PR minimal let's not worry about it here. It seems like we can set `devicePixelRatio`, but not in `browserType.launch`, so it requires some further investigation as to how to do this with WebKit: https://playwright.dev/docs/api/class-browsertype#browser-type-launch-persistent-context

**Comment by tbiethman** [comment]
Code looks good in this initial state. Here are a couple functional findings that don't necessarily need to hold up this initial PR:

macOS
* `Focus` button in launchpad does not focus Webkit browser.
* I'm seeing GraphQL errors that are only present within the Webkit browser. Both e2e/component testing projects seem to be effected, but it can take longer for the error to be presented in component tests.
Screen Shot 2022-08-10 at 10 03 36 AM

* Using the Tab key only seems to focus certain elements within the Webkit browser. I imagine this is related to the Safari setting to "Press Tab to highlight each item on a webpage", which must be enabled to behave similar to Chrome/Firefox. Not sure if there's a corresponding launch flag?
Screen Shot 2022-08-10 at 10 29 26 AM

Windows
* Seeing various failures in both e2e and component tests in our internal projects. I'm sure there are various incompatibilities we'll want to document, not sure if these are known about or not:
Screen Shot 2022-08-10 at 10 57 09 AM
Screen Shot 2022-08-10 at 10 58 01 AM

**Review by tbiethman — APPROVED** [review]

**Comment by flotwig** [comment]
@tbiethman Thanks for the detailed feedback!

>     * `Focus` button in launchpad does not focus Webkit browser.

Marked it as "less critical" here: https://github.com/cypress-io/cypress/issues/6422 I'm using `page.bringToFront` to focus but it seems to not behave the same as Chrome.

>     * I'm seeing GraphQL errors that are only present within the Webkit browser. Both e2e/component testing projects seem to be effected, but it can take longer for the error to be presented in component tests.

Added as a must-fix to #6422

>     * Using the Tab key only seems to focus certain elements within the Webkit browser. I imagine this is related to the Safari setting to "Press Tab to highlight each item on a webpage", which must be enabled to behave similar to Chrome/Firefox. Not sure if there's a corresponding launch flag (or if this has any impact on our `.focus()` command)?

I have noticed some failures in the driver tests related to focus, thanks for the tip on this. I'll keep this in mind when looking at those.

>     * Seeing various failures in both e2e and component tests in our internal projects. I'm sure there are incompatibilities we'll want to document, not sure if these are known about or not:

Not known to me yet - I've been working on fixing the driver tests first before checking out other projects, since that should resolve most common issues. I plan on also adding certain system-tests. Limitations that come from those will be documented.

File: packages/app/src/runner/index.ts
Comment (lmiller1990) [inline packages/app/src/runner/index.ts:43]: I always assumed socket.io did this automatically (fallback to polling). I guess not - good to know.

File: packages/server/lib/browsers/utils.ts
Comment (lmiller1990) [inline packages/server/lib/browsers/utils.ts:196]:
```suggestion
    if (!result?.groups?.version) return '0'
```

Should be able to use `?` (if you like the more concise style)

File: packages/server/lib/browsers/utils.ts
Comment (lmiller1990) [inline packages/server/lib/browsers/utils.ts:219]: Nit, but should we just use the `major` function from the `semver` package we've already got for these in general? Eg

```js
import semver from 'semver'

const majorVersion = semver.major(version)
```

We could even change `getWebKitBrowserVersion` to return the result of this:


```js
return semver.major(result?.groups?.version) ?? '0.0.0'
```

File: packages/server/lib/browsers/webkit-automation.ts
Comment (lmiller1990) [inline packages/server/lib/browsers/webkit-automation.ts:15]:
```suggestion
const extensionMap = {
  'no_restriction': 'None',
  'lax': 'Lax',
  'strict': 'Strict',
} as const

function convertSameSiteExtensionToCypress (str: CyCookie['sameSite']): 'None' | 'Lax' | 'Strict' | undefined {
  return str ? extensionMap[str] : undefined
}
```

If you want to get rid of the `any`.

File: packages/server/lib/browsers/webkit-automation.ts
Comment (lmiller1990) [inline packages/server/lib/browsers/webkit-automation.ts:48]:
```suggestion
    sameSite: convertSameSiteExtensionToCypress(cookie.sameSite),
```

Seems `!` isn't necessary - at least, no error in my IDE.

File: packages/server/lib/browsers/webkit.ts
Comment (lmiller1990) [inline packages/server/lib/browsers/webkit.ts:57]: I notice we pass in a `resetPage` function to the constructor then use it within a variable we assign in the constructor. Since this function is already webkit specific, is there a compelling reason to pass all this logic as a function?

Eg - should we just move the entire body of `resetPage` ([this part](https://github.com/cypress-io/cypress/pull/15533/files#diff-d79130e57e24b51b5f19ffab413c34c1370b0524a971192d366af18838477ad7R35-R53)) to be declared inside of the `WebkitAutomation` class, and just pass it what's necessary (in this case, that would be the `url` and `pwBrowser`, I guess?)

Just an idea - this logic feels like it belongs to `WebkitAutomation` more than it does in a `browsers/webkit` module.

**Review by lmiller1990 — COMMENTED** [review]
Left some comments, they are mainly related to style, though - still trying the functionality out.

**Comment by lmiller1990** [comment]
I found something strange. I don't think it needs to block this, but if I start run mode, then cancel it mid-run, global mode opens after a few seconds.

It's also kind of slow (I guess this is just how it is, at least for now). Since this is development only, I think it's fine to merge up, I'll keep posting things I notice as I continue to play around with it (which will be easier once it's merged).


https://user-images.githubusercontent.com/19196536/184058318-e7e6cf1a-ff86-415b-b81f-368464cad4ff.mov

**Review by lmiller1990 — APPROVED** [review]
Left some potential improvements to code and noted a weird bug I encountered.

File: packages/app/src/runner/index.ts
Comment (flotwig) [reply packages/app/src/runner/index.ts:43]: Yeah, I thought so too. According to that Gh discussion, it only falls back if the websocket API is completely missing, it doesn't fall back on network errors. In WebKit, the API exists, but the connection itself is failing for some reason.

File: packages/server/lib/browsers/webkit.ts
Comment (flotwig) [reply packages/server/lib/browsers/webkit.ts:57]: Agreed, this became more obvious as I added more per-context setup stuff. I'm refactoring this in https://github.com/cypress-io/cypress/pull/23232

**Comment by flotwig** [comment]
@lmiller1990 I've noticed that too. Haven't been able to track it down yet, but it's super annoying. I have it captured as a must-fix here: https://github.com/cypress-io/cypress/issues/6422

File: packages/server/lib/browsers/webkit-automation.ts
Comment (flotwig) [reply packages/server/lib/browsers/webkit-automation.ts:48]: Playwright's `Cookie` type requires a value for SameSite, although this seems to work fine for the basic purpose of getting Cypress running, there is a test failure related to SameSite that's probably caused by this, which I'll fix in #23232 

I get this error in vscode:

```
Type '"None" | "Lax" | "Strict" | undefined' is not assignable to type '"None" | "Lax" | "Strict"'.
  Type 'undefined' is not assignable to type '"None" | "Lax" | "Strict"'.
```

File: packages/server/lib/browsers/webkit-automation.ts
Comment (flotwig) [reply packages/server/lib/browsers/webkit-automation.ts:21]: Nice, TIL about `as const`. 9ac53fd8fad566d3cc0d4ecaa7ea59c33e66e631

File: packages/server/lib/browsers/utils.ts
Comment (flotwig) [reply packages/server/lib/browsers/utils.ts:219]: I'm a fan of not using dependencies wherever reasonable. AFAIK we only use `semver` in one place in the server as it is.

Comment (lmiller1990) [reply packages/server/lib/browsers/utils.ts:219]: Agree on this general sentiment - if we only use `semver` in one place, maybe we should just drop it (and change that one place to not use it).

**Comment by weyert** [comment]
Any plans to mention my name when release this feature?

**Comment by BlueWinds** [comment]
@weyert  - Traditionally we'd do something like this: https://docs.cypress.io/api/commands/selectFile#Community-Recognition. @flotwig - I don't see that we have a docs PR for this, just a reasonable reminder to make sure we mention community contributions when it's ready!

**Comment by flotwig** [comment]
> Any plans to mention my name when release this feature?

@weyert I was planning on it, since this does use your initial approach of driving PW. I actually meant to have you as a `Co-authored-by` on this PR but I guess I managed to merge without including the PR message in the commit message - oops. I'll credit you in the release announcement and docs, unless you don't want to be recognized.

**Comment by weyert** [comment]
I would greatly appreciate that, if you would do that. Thank you 🤗❤️

**Pull request merged**
